astropy/io/fits/util.py

```diff
@@ -490,12 +502,38 @@ def _fileobj_normalize_mode(f):
             # This shouldn't happen?
             return None
 
+    # I've noticed that sometimes Python can produce modes like 'r+b' which I
+    # would consider kind of a bug--mode strings should be normalized. Let's
+    # normalize it for them:
     if '+' in mode:
         mode = mode.replace('+', '')
         mode += '+'
 
-    if _fileobj_is_append_mode(f) and 'a' not in mode:
-        mode = mode.replace('r', 'a').replace('w', 'a')
+    # There was (or still is) some format that loses or hides if it was
+    # opened in append mode... check it:
+
+    # TODO: This is probably superfluous the old corner cases where FILEIO
+    # objects didn't keep track of their append-flag are over (?)
+    # At least the bugfix was ported to python 2.7 and 3.3+:
+    # http://bugs.python.org/issue18876
+    the_investigative_result_if_in_appendmode = _fileobj_is_append_mode(f)
+
+    # Compare if the investigative result returns the same as 'a' in mode.
+    if the_investigative_result_if_in_appendmode != ('a' in mode):
+        # If the results differ we got some format that passed unnoticed but
+        # does not keep it's append-mode. Issue a warning to see if it's
+        # so that people can report it here. If no reports come in one could
+        # omit these tests in favor of a simple "'a' in mode".
+        warnings.warn(
+            "This should not have happened. If you see this Warning please "
+            "report to https://github.com/astropy/astropy/issues and state "
+            "which kind of object you wanted to open and in which mode. We "
+            "apologize for the inconvenience.")
```

Comment:
Oh, I missed that because I simply copied the old warning. Will fix it with the next commit.